pkcs12/src/kdf.rs

```diff
@@ -0,0 +1,108 @@
+//! Implementation of the key derivation function
+//! [RFC 7292 Appendix B](https://datatracker.ietf.org/doc/html/rfc7292#appendix-B)
+
+use alloc::vec::Vec;
+use digest::{core_api::BlockSizeUser, Digest, FixedOutputReset, OutputSizeUser, Update};
+use zeroize::Zeroize;
+
+/// Transform a utf-8 string in a unicode (utf16) string as binary array.
+/// The Utf16 code points are stored in big endian format with two trailing zero bytes.
+fn str_to_unicode(utf8_str: &str) -> Vec<u8> {
+    let mut utf16_bytes = Vec::new();
+    // reserve max number of required bytes to avoid re-allocation
+    utf16_bytes.reserve(utf8_str.len() * 2 + 2);
+    for code_point in utf8_str.encode_utf16().chain(Some(0)) {
+        utf16_bytes.extend(code_point.to_be_bytes());
+    }
+    utf16_bytes
+}
+
+/// Specify the usage type of the generated key
+/// This allows to derive distinct encryption keys, IVs and MAC from the same password or text
+/// string.
+pub enum Pkcs12KeyType {
+    /// Use key for encryption
+    EncryptionKey = 1,
+    /// Use key as initial vector
+    Iv = 2,
+    /// Use key as MAC
+    Mac = 3,
+}
+
+/// Derives `key` of type `id` from `pass` and `salt` with length `key_len` using `rounds`
+/// iterations of the algorithm
+/// ```rust
+/// let key = pkcs12::kdf::derive_key::<sha2::Sha256>("top-secret", &[0x1, 0x2, 0x3, 0x4],
+///     pkcs12::kdf::Pkcs12KeyType::EncryptionKey, 1000, 32);
+/// ```
+pub fn derive_key<D>(
+    pass: &str,
+    salt: &[u8],
+    id: Pkcs12KeyType,
+    rounds: i32,
+    key_len: usize,
+) -> Vec<u8>
+where
+    D: Digest + FixedOutputReset + BlockSizeUser,
+{
+    let mut digest = D::new();
+    let mut pass_utf16 = str_to_unicode(pass);
+    let output_size = <D as OutputSizeUser>::output_size();
+    let block_size = D::block_size();
+    let slen = block_size * ((salt.len() + block_size - 1) / block_size);
+    let plen = block_size * ((pass_utf16.len() + block_size - 1) / block_size);
+    let ilen = slen + plen;
+    let mut init_key = vec![0u8; ilen];
+    for i in 0..slen {
+        init_key[i] = salt[i % salt.len()];
+    }
+    for i in 0..plen {
+        init_key[slen + i] = pass_utf16[i % pass_utf16.len()];
+    }
+    pass_utf16.zeroize();
```

Comment:
This looks like step2 & 3:
```


   2.  Concatenate copies of the salt together to create a string S of
       length v(ceiling(s/v)) bits (the final copy of the salt may be
       truncated to create S).  Note that if the salt is the empty
       string, then so is S.

   3.  Concatenate copies of the password together to create a string P
       of length v(ceiling(p/v)) bits (the final copy of the password
       may be truncated to create P).  Note that if the password is the
       empty string, then so is P.
 ```
 
 Could you move code around and copy the algorithm from the RFC as comments?

Reply:
I have added comments to the code describing the single steps of the implementation, as far as possible since I derive at some point slightly from the algorithm.